File: API/1650_Simple_Ema_Crossover/CS/SimpleEmaCrossoverStrategy.cs
using System;
using System.Collections.Generic;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Simple EMA crossover strategy.
/// Enters long when the fast EMA crosses above the slow EMA and short when the opposite occurs.
/// </summary>
public class SimpleEmaCrossoverStrategy : Strategy
{
	private readonly StrategyParam<int> _periods;
	private readonly StrategyParam<Unit> _stopLoss;
	private readonly StrategyParam<Unit> _takeProfit;
	private readonly StrategyParam<DataType> _candleType;

	private decimal _prevFast;
	private decimal _prevSlow;

	/// <summary>
	/// Period for the fast EMA.
	/// </summary>
	public int Periods
	{
		get => _periods.Value;
		set => _periods.Value = value;
	}

	/// <summary>
	/// Stop-loss distance.
	/// </summary>
	public Unit StopLoss
	{
		get => _stopLoss.Value;
		set => _stopLoss.Value = value;
	}

	/// <summary>
	/// Take-profit distance.
	/// </summary>
	public Unit TakeProfit
	{
		get => _takeProfit.Value;
		set => _takeProfit.Value = value;
	}

	/// <summary>
	/// Candle type.
	/// </summary>
	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	/// <summary>
	/// Initialize strategy parameters.
	/// </summary>
	public SimpleEmaCrossoverStrategy()
	{
		_periods = Param(nameof(Periods), 17)
			.SetGreaterThanZero()
			.SetDisplay("EMA Period", "Period for the fast EMA", "Indicators")
			.SetCanOptimize(true)
			.SetOptimize(10, 30, 1);

		_stopLoss = Param(nameof(StopLoss), new Unit(31m, UnitTypes.Absolute))
			.SetDisplay("Stop Loss", "Stop-loss distance in price", "Risk")
			.SetCanOptimize(true)
			.SetOptimize(10, 100, 5);

		_takeProfit = Param(nameof(TakeProfit), new Unit(69m, UnitTypes.Absolute))
			.SetDisplay("Take Profit", "Take-profit distance in price", "Risk")
			.SetCanOptimize(true)
			.SetOptimize(10, 100, 5);

		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles for analysis", "General");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();
		_prevFast = default;
		_prevSlow = default;
	}

	/// <inheritdoc />
	protected override void OnStarted(DateTimeOffset time)
	{
		base.OnStarted(time);

		StartProtection(takeProfit: TakeProfit, stopLoss: StopLoss);

		var fastEma = new ExponentialMovingAverage { Length = Periods };
		var slowEma = new ExponentialMovingAverage { Length = Periods + 2 };

		var subscription = SubscribeCandles(CandleType);
		subscription
			.Bind(fastEma, slowEma, ProcessCandle)
			.Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawIndicator(area, fastEma);
			DrawIndicator(area, slowEma);
			DrawOwnTrades(area);
		}
	}

	private void ProcessCandle(ICandleMessage candle, decimal fast, decimal slow)
	{
		if (candle.State != CandleStates.Finished)
			return;

		if (!IsFormedAndOnlineAndAllowTrading())
			return;

		if (_prevFast != default && _prevSlow != default)
		{
			var crossUp = _prevFast < _prevSlow && fast > slow;
			var crossDown = _prevFast > _prevSlow && fast < slow;

			if (crossUp && Position <= 0)
				BuyMarket();
			else if (crossDown && Position >= 0)
				SellMarket();
		}

		_prevFast = fast;
		_prevSlow = slow;
	}
}

File: API/1650_Simple_Ema_Crossover/PY/simple_ema_crossover_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import ExponentialMovingAverage
from StockSharp.Algo.Strategies import Strategy


class simple_ema_crossover_strategy(Strategy):
    def __init__(self):
        super(simple_ema_crossover_strategy, self).__init__()
        self._periods = self.Param("Periods", TimeSpan.FromHours(4)) \
            .SetDisplay("EMA Period", "Period for the fast EMA", "Indicators")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(4))) \
            .SetDisplay("Candle Type", "Type of candles for analysis", "General")
        self._prev_fast = 0.0
        self._prev_slow = 0.0
        self._has_prev = False

    @property
    def periods(self):
        return self._periods.Value

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(simple_ema_crossover_strategy, self).OnReseted()
        self._prev_fast = 0.0
        self._prev_slow = 0.0
        self._has_prev = False

    def OnStarted(self, time):
        super(simple_ema_crossover_strategy, self).OnStarted(time)
        fast_ema = ExponentialMovingAverage()
        fast_ema.Length = self.periods
        slow_ema = ExponentialMovingAverage()
        slow_ema.Length = self.periods
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(fast_ema, slow_ema, self.on_process).Start()
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, fast_ema)
            self.DrawIndicator(area, slow_ema)
            self.DrawOwnTrades(area)

    def on_process(self, candle, fast, slow):
        if candle.State != CandleStates.Finished:
            return
        if self._has_prev:
            cross_up = self._prev_fast < self._prev_slow and fast > slow
            cross_down = self._prev_fast > self._prev_slow and fast < slow
            if cross_up and self.Position <= 0:
                self.BuyMarket()
            elif cross_down and self.Position >= 0:
                self.SellMarket()
        self._prev_fast = fast
        self._prev_slow = slow
        self._has_prev = True

    def CreateClone(self):
        return simple_ema_crossover_strategy()
